cumulusci/tasks/salesforce/promote_2gp_package.py

```diff
@@ -0,0 +1,308 @@
+                subscriber_package_version["SubscriberPackageId"]
+            )
+            package_2_version = self._query_Package2Version(spv_id)
+
+            info = {
+                "name": subscriber_package["Name"],
+                "release_state": subscriber_package_version["ReleaseState"],
+                "version_id": spv_id,
+            }
+            if package_2_version:
+                info["is_2gp"] = True
+                info["is_promoted"] = package_2_version["IsReleased"]
+                info["Package2VersionId"] = package_2_version["Id"]
+            else:
+                # It's a 1GP dependency if no Package2Version record is present
+                info["is_2gp"] = False
+                info["is_promoted"] = None
+                info["Package2VersionId"] = None
+
+            dependencies.append(info)
+
+        self.logger.info(f"Total dependencies found: {len(dependencies)}")
+        return dependencies
+
+    def _get_dependency_spv_ids(self, spv_id: str) -> List[str]:
+        """
+        @param spv_id: SubscriberPackageVersionId to fetch dependencies for
+        @return: list of SubscriberPackageVersionIds (04t) of dependency packages
+        """
+        subscriber_package_version = self._query_SubscriberPackageVersion(spv_id)
+        dependencies = subscriber_package_version["Dependencies"] or {"ids": []}
+        dependencies = [d["subscriberPackageVersionId"] for d in dependencies["ids"]]
+        return dependencies
+
+    def _process_one_gp_deps(self, dependencies: List[Dict]) -> None:
+        """
+        Log any 1GP dependecies that are present.
+
+        @param dependencies: list of dependencies to process
+        """
+        one_gp_deps = self._filter_one_gp_deps(dependencies)
+        if one_gp_deps:
+            self.logger.warning("This package has the following 1GP dependencies:")
+            self.logger.warning("")
+            for dep in one_gp_deps:
+                self.logger.warning(
+                    f"    Package Name: {dep['name']:20} ReleaseState: {dep['release_state']}"
+                )
+
+    def _process_two_gp_deps(self, dependencies: List[Dict]) -> bool:
+        """
+        If we're auto-promoting then auto-promote!
+        Otherwise, log which dependencies need promoting.
+
+        @param dependencies: list of dependencies to process
+        @return: a should_exit indicator to tell _run_task() whether or not to exit
+        """
+        two_gp_deps = self._filter_two_gp_deps(dependencies)
+        unpromoted_two_gp_deps = self._filter_unpromoted_two_gp_deps(dependencies)
+
+        self.logger.info(f"Total 2GP dependencies: {len(two_gp_deps)}")
+        self.logger.info(f"Unpromoted 2GP dependencies: {len(unpromoted_two_gp_deps)}")
+
+        should_exit = False
+        if unpromoted_two_gp_deps and self.options.get("auto_promote", False):
+            [self._promote_2gp_package(d) for d in unpromoted_two_gp_deps]
```

Comment:
Using a list comprehension here caught me off guard because I think of their primary purpose as being to construct a list, not to perform a side effect. Would be a bit easier to read if you write it as a normal for loop on 2 lines, I think.